broken tlogclient multiple server unit test

A tlogclient unit test has failed now a couple of times on master code, log:

```
t=2017-07-07T01:42:14+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:33302" module=global caller=redis.go:47
t=2017-07-07T01:42:14+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:40954" module=global caller=redis.go:47
t=2017-07-07T01:42:14+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:45127" module=global caller=redis.go:47
t=2017-07-07T01:42:14+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:33602" module=global caller=redis.go:47
t=2017-07-07T01:42:14+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:37472" module=global caller=redis.go:47
t=2017-07-07T01:42:14+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:42434" module=global caller=redis.go:47
t=2017-07-07T01:42:14+0000 lvl=dbug msg="received handshake request from incoming connection" module=global caller=server.go:157
t=2017-07-07T01:42:14+0000 lvl=dbug msg="incoming connection checks out with version: 1.1.0-alpha" module=global caller=server.go:167
t=2017-07-07T01:42:14+0000 lvl=dbug msg="create vdisk with expectedSequence:0" module=global caller=vdisk_mgr.go:65
t=2017-07-07T01:42:14+0000 lvl=dbug msg="handshake phase successfully completed" module=global caller=server.go:192
t=2017-07-07T01:42:14+0000 lvl=dbug msg="replying handshake with status: OK" module=global caller=server.go:219
t=2017-07-07T01:42:18+0000 lvl=info msg="tlogclient : reconnect from read because of : EOF" module=global caller=client.go:336
FAIL
FAIL	github.com/zero-os/0-Disk/tlog/tlogclient	56.603s
```

4 `make test`, all of them failed on `gonbdserver` package

```
goroutine 56 [IO wait]:
net.runtime_pollWait(0x7f83b6483f58, 0x72, 0x7)
	/home/ibk/go/src/runtime/netpoll.go:164 +0x59
net.(*pollDesc).wait(0xc42001a0d8, 0x72, 0xae8460, 0xae45f8)
	/home/ibk/go/src/net/fd_poll_runtime.go:75 +0x38
net.(*pollDesc).waitRead(0xc42001a0d8, 0xc4225ca080, 0x10)
	/home/ibk/go/src/net/fd_poll_runtime.go:80 +0x34
net.(*netFD).Read(0xc42001a070, 0xc4225ca080, 0x10, 0x10, 0x0, 0xae8460, 0xae45f8)
	/home/ibk/go/src/net/fd_unix.go:250 +0x1b7
net.(*conn).Read(0xc420190008, 0xc4225ca080, 0x10, 0x10, 0x0, 0x0, 0x0)
	/home/ibk/go/src/net/net.go:181 +0x70
io.ReadAtLeast(0x7f83b6484298, 0xc420190008, 0xc4225ca080, 0x10, 0x10, 0x10, 0xaf0c40, 0x8255e0, 0xc4225ca050)
	/home/ibk/go/src/io/io.go:307 +0xa9
io.ReadFull(0x7f83b6484298, 0xc420190008, 0xc4225ca080, 0x10, 0x10, 0x10, 0x0, 0x0)
	/home/ibk/go/src/io/io.go:325 +0x58
encoding/binary.Read(0x7f83b6484298, 0xc420190008, 0xaede60, 0xb405f8, 0x7c89a0, 0xc4225ca050, 0x0, 0x0)
	/home/ibk/go/src/encoding/binary/binary.go:244 +0x1012
github.com/zero-os/0-Disk/gonbdserver/nbd.(*IntegrityTest).Receiver(0xc4201a40d0)
	/home/ibk/pathgo/src/github.com/zero-os/0-Disk/gonbdserver/nbd/integrity_test.go:275 +0x199
created by github.com/zero-os/0-Disk/gonbdserver/nbd.(*IntegrityTest).Run
	/home/ibk/pathgo/src/github.com/zero-os/0-Disk/gonbdserver/nbd/integrity_test.go:381 +0xd1

goroutine 57 [select]:
github.com/zero-os/0-Disk/gonbdserver/nbd.(*IntegrityTest).Ticker(0xc4201a40d0)
	/home/ibk/pathgo/src/github.com/zero-os/0-Disk/gonbdserver/nbd/integrity_test.go:358 +0x2d7
created by github.com/zero-os/0-Disk/gonbdserver/nbd.(*IntegrityTest).Run
	/home/ibk/pathgo/src/github.com/zero-os/0-Disk/gonbdserver/nbd/integrity_test.go:382 +0xf6
FAIL	github.com/zero-os/0-Disk/gonbdserver/nbd	120.860s
?   	github.com/zero-os/0-Disk/log	[no test files]
?   	github.com/zero-os/0-Disk/nbdserver	[no test files]
ok  	github.com/zero-os/0-Disk/nbdserver/ardb	23.944s
ok  	github.com/zero-os/0-Disk/nbdserver/lba	0.013s
?   	github.com/zero-os/0-Disk/redisstub	[no test files]
ok  	github.com/zero-os/0-Disk/tlog	0.010s
?   	github.com/zero-os/0-Disk/tlog/schema	[no test files]
ok  	github.com/zero-os/0-Disk/tlog/tlogclient	14.821s
?   	github.com/zero-os/0-Disk/tlog/tlogclient/blockbuffer	[no test files]
?   	github.com/zero-os/0-Disk/tlog/tlogclient/decoder	[no test files]
?   	github.com/zero-os/0-Disk/tlog/tlogclient/examples/send_tlog	[no test files]
?   	github.com/zero-os/0-Disk/tlog/tlogclient/examples/walk_history	[no test files]
?   	github.com/zero-os/0-Disk/tlog/tlogclient/player	[no test files]
ok  	github.com/zero-os/0-Disk/tlog/tlogserver	4.419s
?   	github.com/zero-os/0-Disk/tlog/tlogserver/aggmq	[no test files]
ok  	github.com/zero-os/0-Disk/tlog/tlogserver/erasure	0.002s [no tests to run]
ok  	github.com/zero-os/0-Disk/tlog/tlogserver/server	3.360s
?   	github.com/zero-os/0-Disk/tlog/tlogserver/slavesync	[no test files]
?   	github.com/zero-os/0-Disk/zeroctl	[no test files]
?   	github.com/zero-os/0-Disk/zeroctl/cmd	[no test files]
?   	github.com/zero-os/0-Disk/zeroctl/cmd/config	[no test files]
?   	github.com/zero-os/0-Disk/zeroctl/cmd/copyvdisk	[no test files]
?   	github.com/zero-os/0-Disk/zeroctl/cmd/delvdisk	[no test files]
?   	github.com/zero-os/0-Disk/zeroctl/cmd/list	[no test files]
ok  	github.com/zero-os/0-Disk/zeroctl/cmd/restore	0.133s
Makefile:44: recipe for target 'testgo' failed
make: *** [testgo] Error 1
```

@chrisvdg how is your output?

@iwanbk, mine failed on tlogclient as the previously mentioned error

`
t=2017-07-10T11:44:56+0200 lvl=info msg="tlogclient : reconnect from read because of : EOF" module=global caller=client.go:336
FAIL
FAIL    github.com/zero-os/0-Disk/tlog/tlogclient       37.070s
?       github.com/zero-os/0-Disk/tlog/tlogclient/blockbuffer   [no test files]
?       github.com/zero-os/0-Disk/tlog/tlogclient/decoder       [no test files]
?       github.com/zero-os/0-Disk/tlog/tlogclient/examples/send_tlog    [no test files]
?       github.com/zero-os/0-Disk/tlog/tlogclient/examples/walk_history [no test files]
?       github.com/zero-os/0-Disk/tlog/tlogclient/player        [no test files]
ok      github.com/zero-os/0-Disk/tlog/tlogserver       4.480s
?       github.com/zero-os/0-Disk/tlog/tlogserver/aggmq [no test files]
ok      github.com/zero-os/0-Disk/tlog/tlogserver/erasure       0.002s [no tests to run]
ok      github.com/zero-os/0-Disk/tlog/tlogserver/server        3.564s
?       github.com/zero-os/0-Disk/tlog/tlogserver/slavesync     [no test files]
?       github.com/zero-os/0-Disk/zeroctl       [no test files]
?       github.com/zero-os/0-Disk/zeroctl/cmd   [no test files]
?       github.com/zero-os/0-Disk/zeroctl/cmd/config    [no test files]
?       github.com/zero-os/0-Disk/zeroctl/cmd/copyvdisk [no test files]
?       github.com/zero-os/0-Disk/zeroctl/cmd/delvdisk  [no test files]
?       github.com/zero-os/0-Disk/zeroctl/cmd/list      [no test files]
ok      github.com/zero-os/0-Disk/zeroctl/cmd/restore   0.164s
Makefile:44: recipe for target 'testgo' failed
make: *** [testgo] Error 1
`

gonbdserver did not fail for me
`
ok      github.com/zero-os/0-Disk/gonbdserver/nbd       15.904s
`   

I did find this now, scrolling though the output  
`
--- FAIL: TestMultipleServerBasicMasterFailToFlush (25.19s)
        multiple_server_test.go:144: Create tlog servers
        multiple_server_test.go:153: connect client
        multiple_server_test.go:165: write data to server #1
        multiple_server_test.go:194: write data to server #2
        common_test.go:27: testClientWaitSeqFlushed timeout
`

> is this the same error as https://travis-ci.org/zero-os/0-Disk/jobs/250433968?

Scrolling through the travis logs :

```
t=2017-07-05T16:53:19+0000 lvl=info msg="flush of vdiskID=myimg failed: [error during flush idx 0: redis Send failed: no connectionPool exists for index 0 error during flush idx 1: redis Send failed: no connectionPool exists for index 1]" module=global caller=flusher.go:173

t=2017-07-05T16:53:19+0000 lvl=info msg="flush myimg failed: flush vdiskID=myimg failed" module=global caller=vdisk.go:411

--- FAIL: TestMultipleServerResendHotReload (20.11s)
	multiple_server_test.go:144: Create tlog servers
	multiple_server_test.go:153: connect client
	multiple_server_test.go:165: write data to server #1
	multiple_server_test.go:194: write data to server #2
	common_test.go:27: testClientWaitSeqFlushed timeout
```

Seems to be related to when redis is flushed?

It is still failing for me
```
FAIL    github.com/zero-os/0-Disk/tlog/tlogclient       79.919s
?       github.com/zero-os/0-Disk/tlog/tlogclient/blockbuffer   [no test files]
?       github.com/zero-os/0-Disk/tlog/tlogclient/decoder       [no test files]
?       github.com/zero-os/0-Disk/tlog/tlogclient/examples/send_tlog    [no test files]
?       github.com/zero-os/0-Disk/tlog/tlogclient/examples/walk_history [no test files]
?       github.com/zero-os/0-Disk/tlog/tlogclient/player        [no test files]
ok      github.com/zero-os/0-Disk/tlog/tlogserver       4.429s
?       github.com/zero-os/0-Disk/tlog/tlogserver/aggmq [no test files]
ok      github.com/zero-os/0-Disk/tlog/tlogserver/erasure       0.005s [no tests to run]
ok      github.com/zero-os/0-Disk/tlog/tlogserver/server        3.535s
?       github.com/zero-os/0-Disk/tlog/tlogserver/slavesync     [no test files]
?       github.com/zero-os/0-Disk/zeroctl       [no test files]
?       github.com/zero-os/0-Disk/zeroctl/cmd   [no test files]
?       github.com/zero-os/0-Disk/zeroctl/cmd/config    [no test files]
?       github.com/zero-os/0-Disk/zeroctl/cmd/copyvdisk [no test files]
?       github.com/zero-os/0-Disk/zeroctl/cmd/delvdisk  [no test files]
?       github.com/zero-os/0-Disk/zeroctl/cmd/list      [no test files]
ok      github.com/zero-os/0-Disk/zeroctl/cmd/restore   0.162s
Makefile:44: recipe for target 'testgo' failed
make: *** [testgo] Error 1
```

Same fail message as well:
```
--- FAIL: TestMultipleServerBasicMasterFailToFlush (68.07s)
        multiple_server_test.go:144: Create tlog servers
        multiple_server_test.go:153: connect client
        multiple_server_test.go:165: write data to server #1
        multiple_server_test.go:194: write data to server #2
        common_test.go:27: testClientWaitSeqFlushed timeout
```

--- a/tlog/tlogclient/blockbuffer/buffer.go
+++ b/tlog/tlogclient/blockbuffer/buffer.go
@@ -3,7 +3,6 @@ package blockbuffer
 import (
 	"context"
 	"errors"
-	"math"
 	"sort"
 	"sync"
 	"time"
@@ -22,7 +21,10 @@ var (
 type Buffer struct {
 	lock sync.RWMutex
 
-	// map of sent block
+	// map of sent blocks.
+	// blocks in this map is going to moved to
+	// waitToFlush map after it receive sent confirmation
+	// from the tlog server
 	entries map[uint64]*entry
 
 	// Array of ordered sent sequence.
@@ -41,6 +43,9 @@ type Buffer struct {
 
 	// the highest sequence we received
 	highestSequence uint64
+
+	// last flushed sequence
+	lastFlush uint64
 }
 
 // NewBuffer creates a new tlog blocks buffer
@@ -53,7 +58,7 @@ func NewBuffer(timeout time.Duration) *Buffer {
 	}
 }
 
-// implement sort interface for buffer.seqToTimeout
+// Uint64Slice implements sort interface for buffer.seqToTimeout
 type Uint64Slice []uint64
 
 func (us Uint64Slice) Len() int           { return len(us) }
@@ -66,6 +71,7 @@ func (b *Buffer) SetResendAll() {
 	b.lock.RLock()
 	defer b.lock.RUnlock()
 
+	// empty the seqToTimeout slice
 	b.seqToTimeout = make([]uint64, 0, len(b.entries)+len(b.waitToFlush))
 
 	for seq, ent := range b.entries {
@@ -92,6 +98,10 @@ func (b *Buffer) Add(block *schema.TlogBlock) {
 
 	seq := block.Sequence()
 
+	if seq < b.lastFlush {
+		return
+	}
+
 	// add to array of ordered timeout
 	b.seqToTimeout = append(b.seqToTimeout, seq)
 
@@ -112,18 +122,32 @@ func (b *Buffer) Add(block *schema.TlogBlock) {
 	b.entries[seq] = ent
 }
 
-func (b *Buffer) LenWaitFlush() {
-}
-
 // SetFlushed set this sequence as succesfully flushed
 func (b *Buffer) SetFlushed(seqs []uint64) {
 	b.lock.Lock()
 	defer b.lock.Unlock()
 
+	if len(seqs) == 0 {
+		return
+	}
+
 	for _, seq := range seqs {
 		delete(b.waitToFlush, seq)
-		delete(b.entries, seq) // we normally don't need to do this
+		delete(b.entries, seq)
 	}
+
+	// update lastFlush
+	lastSeq := seqs[len(seqs)-1]
+	if b.lastFlush < lastSeq {
+		b.lastFlush = lastSeq
+	}
+}
+
+// LastFlushed returns last sequence flushed
+func (b *Buffer) LastFlushed() uint64 {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+	return b.lastFlush
 }
 
 // SetSent set this sequence as succesfully sent
@@ -140,7 +164,7 @@ func (b *Buffer) SetSent(seq uint64) {
 	delete(b.entries, seq)
 }
 
-// Returns true this seq need to be re-send
+// NeedResend returns true if this sequence need to be re-send
 func (b *Buffer) NeedResend(seq uint64) bool {
 	b.lock.RLock()
 	defer b.lock.RUnlock()
@@ -209,34 +233,4 @@ func (b *Buffer) TimedOut(ctx context.Context) <-chan *schema.TlogBlock {
 	}()
 
 	return b.readyChan
-}
-
-// MinSequence returns min sequence that this
-// buffer will/currently has.
-func (b *Buffer) MinSequence() uint64 {
-	b.lock.RLock()
-	defer b.lock.RUnlock()
-
-	minUnflushed := b.minSeq(b.waitToFlush)
-	minNotDelivered := b.minSeq(b.entries)
-
-	if minUnflushed < minNotDelivered {
-		return minUnflushed
-	}
-	return minNotDelivered
-}
-
-func (b *Buffer) minSeq(m map[uint64]*entry) uint64 {
-	if len(m) == 0 {
-		return b.highestSequence + 1
-	}
-
-	var min uint64 = math.MaxUint64
-	for seq := range m {
-		if seq < min {
-			min = seq
-		}
-	}
-	return min
-
 }

--- a/tlog/tlogclient/client.go
+++ b/tlog/tlogclient/client.go
@@ -170,7 +170,7 @@ func (c *Client) reconnect(closedTime time.Time) error {
 		// try other server
 		c.shiftServerAddr()
 
-		if err = c.connect(c.blockBuffer.MinSequence(), false); err == nil {
+		if err = c.connect(c.blockBuffer.LastFlushed()+1, false); err == nil {
 			c.lastConnected = time.Now()
 
 			// if reconnect success, sent all unflushed blocks

